### PR TITLE
Migrate SelfAttention and CrossAttention off of ModuleSpec

### PR DESCRIPTION
**Pull request by nschank**
# What does this PR do ?

Demonstrates what a migration away from ModuleSpec would look like. Draft is mostly for demonstration - can break it up or target the first updates differently depending on request.

Associated design doc: https://docs.google.com/document/d/1shyv0iKEzRdevLOlouF_NktbdJazvWifqxUwPXFigQE/edit?tab=t.0#heading=h.uwes2zo47yg6

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [X] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by Skylion007** [comment]
FYI, this typing utility might be useful. I need to clean it up a bit though: https://github.com/pytorch/pytorch/pull/163418/files let's you copy signatures form super classes and validate it.

File: examples/multimodal/layer_specs.py [hunk 1]
```diff
@@ -122,12 +124,16 @@ def get_layer_spec_te(is_vit=False, padding=False) -> ModuleSpec:
...
     )
 
+
 def get_mamba_layer_spec_te(padding=False) -> ModuleSpec:
     attn_mask_type = AttnMaskType.causal
     # Padding mask is needed for e.g. Context Parallel.
     if padding:
         attn_mask_type = AttnMaskType.padding_causal
 
+    assert TELayerNormColumnParallelLinear is not None
```
Comment (Skylion007) [inline examples/multimodal/layer_specs.py:134]: Creating a not_none typing utility (or just using the one in PyTorch) might make your lives a bit easier.

File: megatron/core/extensions/transformer_engine_spec_provider.py [hunk 2]
```diff
@@ -33,19 +33,21 @@ def column_parallel_linear(self) -> type:
...
 
-    def column_parallel_layer_norm_linear(self) -> Optional[type]:
+    def column_parallel_layer_norm_linear(self) -> type[TELayerNormColumnParallelLinear]:
         """Which module for sequential layernorm and linear"""
         return TELayerNormColumnParallelLinear
 
-    def layer_norm(self, rms_norm: bool = False, for_qk: bool = False) -> type:
+    def layer_norm(
+        self, rms_norm: bool = False, for_qk: bool = False
+    ) -> Union[type[FusedLayerNorm], type[TENorm]]:
```
Comment (Skylion007) [inline megatron/core/extensions/transformer_engine_spec_provider.py:50]: Yikes, have a TypeAlias here that references all known LayerNorm classes here if you aren't going to have a parent class.

File: megatron/core/models/backends.py [hunk 3]
```diff
@@ -89,19 +89,21 @@ def column_parallel_linear(self) -> type:
...
 
-    def column_parallel_layer_norm_linear(self) -> Optional[type]:
+    def column_parallel_layer_norm_linear(self) -> None:
         """Which module for sequential layernorm and linear"""
         return None
 
-    def layer_norm(self, rms_norm: bool = False, for_qk: bool = False) -> type:
+    def layer_norm(
+        self, rms_norm: bool = False, for_qk: bool = False
+    ) -> Union[type['FusedLayerNorm'], type[WrappedTorchNorm]]:
```
Comment (Skylion007) [inline megatron/core/models/backends.py:106]: Import this other class Is it defined? Why is it stringified is it defined here too? Type checker is not gooing to like this and fall back to Any, wont it?

File: megatron/core/models/gpt/heterogeneous/heterogeneous_layer_specs.py [hunk 4]
```diff
@@ -106,13 +109,27 @@ def _get_heterogenous_attention_spec(
         )
     else:
         ln = _get_qk_layernorm(use_te, normalization) if qk_layernorm else IdentityOp
+        if use_te:
+            assert TELayerNormColumnParallelLinear is not None
+            assert TEDotProductAttention is not None
+            assert TERowParallelLinear is not None
+            linear_qkv = TELayerNormColumnParallelLinear
+            core_attention = TEDotProductAttention
+            linear_proj = TERowParallelLinear
```
Comment (Skylion007) [inline megatron/core/models/gpt/heterogeneous/heterogeneous_layer_specs.py:118]:
```suggestion
            linear_qkv = not_none(TELayerNormColumnParallelLinear)
            core_attention = not_none(TEDotProductAttention)
            linear_proj = not_none(TERowParallelLinear)

File: megatron/core/models/gpt/heterogeneous/heterogeneous_layer_specs.py [hunk 5]
```diff
@@ -106,13 +109,27 @@ def _get_heterogenous_attention_spec(
...
+
+        else:
+            linear_qkv = ColumnParallelLinear
+            core_attention = DotProductAttention
+            linear_proj = RowParallelLinear
```
Comment (Skylion007) [inline megatron/core/models/gpt/heterogeneous/heterogeneous_layer_specs.py:123]: Use not_none from torch typing utilities or copy it here
```suggestion
        linear_qkv = not_none(TELayerNormColumnParallelLinear)
        core_attention = not_none(TEDotProductAttention)
        linear_proj = not_none(TERowParallelLinear)

File: megatron/core/transformer/attention.py [hunk 6]
```diff
@@ -107,17 +108,143 @@
...
+        self,
+        input_size: int,
+        output_size: int,
+        /,
+        *,
+        config: TransformerConfig,
+        init_method: Callable[[Tensor], None],
+        gather_output: bool,
+        bias: bool,
+        skip_bias_add: bool,
```
Comment (Skylion007) [inline megatron/core/transformer/attention.py:130]: boolean trap, bias coould be an enum that had a skip_bias_add_option

File: megatron/core/transformer/spec_utils.py [hunk 7]
```diff
@@ -24,7 +24,16 @@ class ModuleSpec:
...
+    submodules: object = None
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        """Builds an instance of the module from the spec.
+
+        Args:
+            *args: Positional arguments to be passed to the module init.
+            **kwargs: Keyword arguments to be passed to the module init.
+        """
+        return build_module(self, *args, **kwargs)
```
Comment (Skylion007) [inline megatron/core/transformer/spec_utils.py:36]: Just copy the signature from build_module or do:
```suggestion
        __call__ = build_module

File: megatron/core/typed_torch.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+R_co = TypeVar('R_co', covariant=True)
+
+
+class _Module(Generic[P, R_co], Protocol):
+    """Protocol allowing us to unwrap `forward`."""
+
+    def forward(self, *args: P.args, **kwargs: P.kwargs) -> R_co: ...
+
+
+def apply_module(m: _Module[P, R_co], *, check_subclass: bool = True) -> Callable[P, R_co]:
```
Comment (Skylion007) [inline megatron/core/typed_torch.py:18]: You can use a typing overload here to have check_subclass=True be a static assertion here. Now if check_subclass is True it will require it to follow the _TrueModule Protocol and complain if it doesn't. Mypy, pyright, etc prefer the more specific overload so this should work.

```suggestion
class _TrueModule(torch.nn.Module, _Module, Protocol):
    pass

@typing.overload
def apply_module(m: _TrueModule[P, R_co], *, check_subclass: Literal[True] = True) -> Callable[P, R_co]:
    pass
    
@typing.overload
def apply_module(m: _Module[P, R_co], *, check_subclass: Literal[True] = True) -> typing.Never:
    pass
```

**Comment by nschank** [comment]
@Skylion007 I just saw this, sorry - I was mostly sending this PR as an example while the doc was being discussed. I broke out the first few commits into a prep PR: https://github.com/NVIDIA/Megatron-LM/pull/2668 happy to discuss further there!

File: megatron/core/typed_torch.py (hunk 8 above)
Comment (nschank) [reply megatron/core/typed_torch.py:18]: This is a good idea, I'll add this!

File: megatron/core/models/backends.py (hunk 3 above)
Comment (nschank) [reply megatron/core/models/backends.py:106]: Type checker doesn't care about strings vs. not - I stringified because if the apex import fails, then this type is unbound.

However, this is all more of a draft PR just kind of showing the broader ideas here - I'd like to clean up the imports instead when I actually update this, to avoid needing to be weird like this.

File: examples/multimodal/layer_specs.py (hunk 1 above)
Comment (nschank) [reply examples/multimodal/layer_specs.py:134]: Yeah that'd be fine! I don't actually want to do it this way (it's clearly sorta ugly), it was just a cheap way to get the type-checker to pass locally while I was drafting. Sorry for the confusion :)

File: megatron/core/extensions/transformer_engine_spec_provider.py (hunk 2 above)
Comment (nschank) [reply megatron/core/extensions/transformer_engine_spec_provider.py:50]: Ah so this is something that will take some thinking when we get to it - I'm taking advantage of covariance here and simply having the subclass specify what it would actually return, specifically because it allows for the type-checker to actually examine the underlying types for the correct signatures in practice. This is not a desired end-state, but was needed (at least while I was playing around) to actually ensure type-checking occurred at all the call-sites where implementations and `Protocols` needed to be matched up.

In reality, we wouldn't want the `BackendProvider` interface to specify a base class or a union of classes (which would prevent users from providing custom implementations without subclassing), we would want it to actually have methods which return the specific `Protocol` that needs to be satisfied; but this would require a bit more of a refactor since each individual method right now probably needs to be split up in order to describe the different ways in which the layers are used. For the moment, since there are only a few `BackendProviders` and they are mostly used non-polymorphically, it was cheaper to simply push the subclasses to declare the exact types they were returning, since that made ~almost all callers then type-check that the return types actually satisfy the `Protocols`.

In coming PRs, expect this to be done more cleanly

File: megatron/core/transformer/attention.py (hunk 6 above)
Comment (nschank) [reply megatron/core/transformer/attention.py:130]: I am explicitly not going to try to fix the existing interfaces in these PRs - these `Protocols` will be exact replicas of the current way in which the underlying modules will be used, to ensure backwards compatibility. (I'm happy to _also_ help fix things, but I don't want to get caught in the trap of doing both at once and confusing everyone)

File: megatron/core/transformer/spec_utils.py (hunk 7 above)
Comment (nschank) [reply megatron/core/transformer/spec_utils.py:36]: What do you mean 'copy the signature'? Happy to do it either way since they're equivalent, but the existing methods has no type hints or docstring!

**Comment by nschank** [comment]
> FYI, this typing utility might be useful. I need to clean it up a bit though: https://github.com/pytorch/pytorch/pull/163418/files let's you copy signatures form super classes and validate it.

You're doing god's work there, but unfortunately that wouldn't help me in the short-term - until all existing modules (both inside and outside of Megatron, plus all custom modules provided by users) are consistently annotated correctly, I will still need a way to figure out the correct signature of a module nonintrusively. I view the typed_torch library as a necessary evil which can be discarded easily when the time comes!

File: megatron/core/typed_torch.py (hunk 8 above)
Comment (nschank) [reply megatron/core/typed_torch.py:18]: Oh wait no I tried this when I originally wrote it - `Protocols` cannot extend non-`Protocol` classes. Unfortunately python typing is fairly limited and doesn't allow for higher-kinded or constrained types :(
